File: generate/reference/duality-of-compilation/ocaml/match_options.ml
let rec attempt i =
  if i = 0 then
    Some i
  else
    match attempt (i - 1) with
    | Some x -> Some (x + 1)
    | None -> None

let main n =
  match attempt n with
  | Some x -> x
  | None -> -1

let () =
  match Sys.argv with
  | [| _ |] -> print_endline (string_of_int (main 10))
  | [| _; s |] -> (
      match int_of_string_opt s with
      | Some n -> print_endline (string_of_int (main n))
      | None -> print_endline "argument is not a number")
  | _ -> print_endline "arguments are too many"

File: generate/reference/duality-of-compilation/rust/match_options.rs
use std::env;

fn attempt(i: isize) -> Option<isize> {
    if i == 0 {
        Some(i)
    } else {
        match attempt(i - 1) {
            Some(x) => Some(x + 1),
            None => None,
        }
    }
}

fn run(n: isize) -> isize {
    match attempt(n) {
        Some(x) => x,
        None => -1,
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let arg = match args.get(1) {
        None => 10,
        Some(val) => match val.parse::<isize>() {
            Ok(val) => val,
            Err(_) => {
                println!("argument is not a number");
                return;
            }
        },
    };

    println!("{}", run(arg));
}
